### ISSUE TITLE
Dropout of attn_weights

### ISSUE THREAD

**Issue by arendu** [pytorch/fairseq#351]
The [implementation](https://github.com/pytorch/fairseq/blob/b1521f962e4ca670311c0cd0c8b1dadf310cb242/fairseq/modules/multihead_attention.py#L176) of the transformer seems to apply a dropout after the `attn_weights` have been normalized, which would result in the final sum of `attn_weights != 1.0`. 

A better way to ensure the final attn_weights sum to 1 could be to do something like:
```
attn_weights = F.dropout(attn_weights, p=self.dropout, training=self.training)
attn_weights[attn_weights==0] = float(-inf)
attn_weights = F.softmax(attn_weights.float(), dim=-1).type_as(attn_weights)
```

**Comment by myleott** [comment]
I believe during training Dropout scales the outputs by 1/(1-p), so it should still sum to 1. See: https://pytorch.org/docs/stable/nn.html#torch.nn.Dropout

**Comment by myleott** [comment]
Edit: actually you’re right, good catch! Dropout scales the output by 1/(1-p), but there’s no guarantee that this sums to 1.

Note that tensor2tensor seems to do the same thing as well: https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/common_attention.py#L1473

**Comment by myleott** [comment]
Closing since this matches the reference implementation. Happy to accept a PR that fixes it in a backward compatible way! (may be a bit tricky, but possible with upgrade_state_dict)
